### ISSUE TITLE
TF_AddGradientsWithPrefix doesn't seem to lock the gradients properly

### ISSUE THREAD

**Issue by Craigacp** [tensorflow/tensorflow#48855]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Using TF-Java
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS
- TensorFlow installed from (source or binary): Maven Central binary
- TensorFlow version (use command below): v2.4.1

**Describe the current behavior**

When training a model I'm hitting a non-determinism issue, the models are initialized identically, and fed identical data in an identical order, with inter-op and intra-op parallelism both set to 1, but I can get complete training failure when building simple MLPs and CNNs on MNIST.

TF-Java uses TF_AddGradientsWithPrefix to construct the gradients. We then wrap that in Optimizers to make a higher level front end to training, similar to the v1 optimizers package, and Keras in v2. As gradients are only available in Graph mode in the C API our code is most similar to the v1 optimizers package. I think we want to have the gating behaviour set to `GATE_OP` (https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Optimizer), but that's all implemented entirely in python, and the C API codepath doesn't seem to add the required control dependencies to ensure the gradients are all computed before being back-propped any further (in Python that's performed here - https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/ops/gradients_util.py#L692, but there doesn't seem to be an equivalent in the C API here - https://github.com/tensorflow/tensorflow/blob/5dcfc51118817f27fad5246812d83e5dccdc5f72/tensorflow/cc/framework/gradients.cc#L568).

**Describe the expected behavior**

The model should train identically with identical gradients.

**Standalone code to reproduce the issue**

See this test in TF-Java - https://github.com/Craigacp/tensorflow-java/blob/03402b30e271a46438c1feff12409095737f3d3e/tensorflow-framework/src/test/java/org/tensorflow/framework/optimizers/GradientDescentTest.java#L121. This test when run complains that out of the 20 runs some of them diverged even when trained using identical graph defs on identical data, for an identical number of steps, when using single threaded computation.

**Comment by Craigacp** [comment]
Has anyone had a chance to look at this?

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by Craigacp** [comment]
It is still broken, though we've found a work around in Java which induces better locking of the gradients. It's not a proper fix, and so it would be good to have this be fixed in the C++ implementation.

**Comment by Kayyuri** [comment]
Hi @Craigacp,

Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
